### PR TITLE
Backend category performance fix

### LINKED ISSUE #34109: Bug on category count (performance issue)
In src/vendor/magento/module-catalog/Model/ResourceModel/Category/Collection.php

Line 330-333 :

```
                $productsCount = isset($categoriesProductsCount[$item->getId()])
                    ? (int)$categoryProductsCount[$item->getId()]
                    : $this->getProductsCountFromCategoryTable($item, $websiteId);
```

should be 

```
                $productsCount = isset($categoryProductsCount[$item->getId()])
                    ? (int)$categoryProductsCount[$item->getId()]
                    : $this->getProductsCountFromCategoryTable($item, $websiteId);
```

( categoriesProductsCount instead of categoryProductsCount )

This is causing performance issue on backend category page especially when having a big amount of categories/products.

### PR DESCRIPTION
**Pull request by Nuranto**
### Description (*)

- Add product-category index table for default store (used for backend category page - product count)
- Use index table in backend category page for product count

### Fixed Issues (if relevant)

1. Fixes magento/magento2#34109 

### Manual testing scenarios (*)

1. Measure time (or count SQL requests, in my case 14K+) of /admin/catalog/category/ and /admin/catalog/category/store/X pages
2. Pull changees
3. Reindex catalog_category_product
4. Measure again (or count again, in my care 7K~)

### Questions or comments

To make the change work on default store (admin store), I had to enable reindexing catalog_product_category on default store.
Which mean there is a new table created and indexed : catalog_category_product_index_store0 (and catalog_category_product_index_store0_replica)

In Collection class, you see I use ObjectManager instead of passing class through constructor. I did that to avoid break classes that could extends this one (third par modules). But if you want, I can change that, I don't know what your policy is about that.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by Nuranto** [comment]
@magento run all tests

File: app/code/Magento/Catalog/Model/Indexer/Category/Product/AbstractAction.php [hunk 1]
```diff
@@ -343,6 +334,19 @@ protected function getNonAnchorCategoriesSelect(Store $store)
...
+            if ($store->getId() != 0) {
+                $select->joinInner(
+                    ['cpw' => $this->getTable('catalog_product_website')],
+                    'cpw.product_id = ccp.product_id',
+                    []
+                )->where(
+                    'cpw.website_id = ?',
+                    $store->getWebsiteId()
+                )->where(
+                    'cc.path LIKE ' . $this->connection->quote($rootPath . '/%')
```
Comment (Bashev) [inline app/code/Magento/Catalog/Model/Indexer/Category/Product/AbstractAction.php:346]: my suggestion is to use regex instead like, to improve performance of the query.

Comment (Nuranto) [reply app/code/Magento/Catalog/Model/Indexer/Category/Product/AbstractAction.php:346]: I thought like was faster than regex ?!

**Review by orlangur — CHANGES_REQUESTED** [review]
`$this->storeManager->getStores(true)` change is quite suspicious to me.

Also, I believe it was intended to not create index tables for admin store.

Count of SQL queries means nothing in terms of performance, how much is that in terms of milliseconds?

**Comment by Nuranto** [comment]
@orlangur I'm sure it was intended. However when they did, they forgot about category backend page. 

My changes in `app/code/Magento/Catalog/Model/ResourceModel/Category/Collection.php` confirms that. That code was totally unused as it was. (wrong variable name, and even if it was correct, it was looking into the deprecated catalog_category_product_index table => that's the reason of other changes.).

Before my changes, it went through `getProductsCountFromCategoryTable` for each category. So 1 sql query for each. In my case, that's 7 780 requests that could be avoided, for one single page load... One of these SQL request was around 0.25ms, so almost 2s gain on TTFB...

However, those time were calculated on local environment, so the gain is probably a bit less on production environments. 

In any case, even if you don't want to add index on admin store, the code `app/code/Magento/Catalog/Model/ResourceModel/Category/Collection.php` still needs to be adapted accordingly..

I'll leave it into your hands

File: app/code/Magento/Catalog/Model/ResourceModel/Category/Collection.php [hunk 2]
```diff
@@ -327,7 +328,7 @@ public function loadProductCount($items, $countRegular = true, $countAnchor = tr
             $countSelect = $this->getProductsCountQuery($categoryIds, (bool)$websiteId);
             $categoryProductsCount = $this->_conn->fetchPairs($countSelect);
             foreach ($anchor as $item) {
-                $productsCount = isset($categoriesProductsCount[$item->getId()])
+                $productsCount = isset($categoryProductsCount[$item->getId()])
```
Comment (orlangur) [inline app/code/Magento/Catalog/Model/ResourceModel/Category/Collection.php:331]: This change which was mentioned in initial ticket sure makes sense.
Could you please improve it even more with

```
$productsCount = (int)($categoryProductsCount[$item->getId()] ?? $this->getProductsCountFromCategoryTable($item, $websiteId));
```

**Review by orlangur — CHANGES_REQUESTED** [review]
@Nuranto changing so much indexers and adding index tables for store=0 seems quite risky to me.

Could you just minimize changes to initial observation with variable name misprint? Will this improve performane flaw much?

File: app/code/Magento/Catalog/Model/ResourceModel/Category/Collection.php (hunk 2 above)
Comment (Nuranto) [reply app/code/Magento/Catalog/Model/ResourceModel/Category/Collection.php:331]: That's what I tried first. But as I was trying to explain, `getProductsCountQuery` (which is supposed to fill `$categoryProductsCount`) try to get data from `catalog_category_product_index`, which is an obsolete table which stays empty... That table has been replaced by `catalog_category_product_index_storeX`, leaving no more general index table for admin...

If you keep only the change in Collection class, so the variable rename, and the use of `$tableMaintainer` in `getProductsCountQuery` method, then we'll have a partial fix that solves the issue when user selects a store view in backend category page. But the issue will remains on default view. That's why I added mechanics for `catalog_category_product_index_store0`...

Comment (Nuranto) [reply app/code/Magento/Catalog/Model/ResourceModel/Category/Collection.php:331]: @slopukhov I see you're the one who made that change (a long time ago ^^).. Maybe you have an opinion on this and a better solution that the one I proposed ?

**Pull request closed without merging**